hudi-client/src/main/java/org/apache/hudi/io/storage/HoodieFileWriterFactory.java

```diff
@@ -34,29 +34,28 @@
 
 import java.io.IOException;
 
-import static org.apache.hudi.common.model.HoodieFileFormat.HOODIE_LOG;
 import static org.apache.hudi.common.model.HoodieFileFormat.PARQUET;
+import static org.apache.hudi.common.model.HoodieFileFormat.HFILE;
 
-public class HoodieStorageWriterFactory {
+public class HoodieFileWriterFactory {
 
-  public static <T extends HoodieRecordPayload, R extends IndexedRecord> HoodieStorageWriter<R> getStorageWriter(
+  public static <T extends HoodieRecordPayload, R extends IndexedRecord> HoodieFileWriter<R> getFileWriter(
```

Comment:
I have created the new HoodieSortedMergeHandle which extends the HoodieMergeHandle. This way, the sort functionality is moved to the HoodieSortedMergeHandle. The correct handle will be initialized and used within CommitActionExecutor.

This was cleaner than an external merge Algorithm to be plugged into the HoodieMergeHandle because:
1.  Sort in MergeHandle is closely tied to reading and writing the records. The algorithm is better suited to cases where all records are already available in memory.
2. Cleaner code as the intention is evident from the handle name.
3. Removes code duplication
4. No change for non-hfile base formats (the existing HoodieMergeHandle will be used).

Please take a look.

Reply:
Regarding AbstractFactoryPattern:

I can group the two reader -writer factories into one single factory:

class HoodieFIleFormatFactory:
    public getFileWriter()
    public getFileReader()

    static HoodieFIleFormatFactory getFileFormatFactory()

IMO, this does not change much but makes it a bit complicated as:
1. File writers require more configuration (e.g. compression algorithm etc)
2. File readers only require file path

In case only a reader is required - initializing a common factory class will require passing parameters (common with writer).

Reply:
I did go thru the AbstractFactoryPattern. Is this what you are insinuating @bvaradar 

```
class FileStorageFactoryProducer {

	FileStorageFactory getFileStorageFactory(Config config){
		if(config.getFileStorageType() == PARQUET) {
			return new ParquetFileStorageFactory(....)
		} else if(config.getFileStorageType() == HFile) {
			return new HFileFileStorageFactory(....)
		} else {
		   // throw exception
		}
	}  
}


interface FileStorageFactory {
	
	public static <T extends HoodieRecordPayload, R extends IndexedRecord> HoodieFileWriter<R> getFileWriter(String instantTime, 
Path path, HoodieTable<T> hoodieTable, HoodieWriteConfig config, Schema schema, 
SparkTaskContextSupplier sparkTaskContextSupplier) throws IOException;

        public static <T extends HoodieRecordPayload, R extends IndexedRecord> HoodieFileReader<R> getFileReader(
Configuration conf, Path path) throws IOException;

}


class ParquetFileStorageFactory implements FileStorageFactory {
	
		// exposes Writer and Reader for Parquet

}


class HFileFileStorageFactory implements FileStorageFactory {
		// exposes Writer and Reader for HFile
}
```